File: Gui/Linn/Gui/SceneGraph/NodeHit.cs
using Linn.Gui.Scenegraph;
using System.Xml;
using Linn.Gui.Resources;
using System;
using System.Globalization;
using System.Drawing;

namespace Linn {
namespace Gui {
namespace Scenegraph {
    
public class NodeHit : Node
{
    public NodeHit() : base(UniqueName("NodeHit")) {
    }
    
    public NodeHit(string aName) : base(aName) {
    }
    
    public override Object Clone() {
        NodeHit h = new NodeHit();
        Clone(h);
        return h;
    }
    
    protected void Clone(NodeHit aNodeHit) {
        Clone((Node)aNodeHit);
        aNodeHit.iWidth = iWidth;
        aNodeHit.iHeight = iHeight;
        aNodeHit.iAllowHits = iAllowHits;
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        NumberFormatInfo nfi = new NumberFormatInfo();
        nfi.NumberDecimalSeparator = ".";
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("Width");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iWidth = float.Parse(list[0].FirstChild.Value, nfi);
        }
        
        list = aXmlNode.SelectNodes("Height");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iHeight = float.Parse(list[0].FirstChild.Value, nfi);
        }
        
        list = aXmlNode.SelectNodes("AllowHits");
        if(list != null) {
            if(list.Count > 0) {
                Assert.Check(list.Count == 1);
                iAllowHits = bool.Parse(list[0].FirstChild.Value);
            }
        }
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        
        NumberFormatInfo nfi = new NumberFormatInfo();
        nfi.NumberDecimalSeparator = ".";
                    
        aWriter.WriteStartElement("Width");
        aWriter.WriteString(iWidth.ToString(nfi));
        aWriter.WriteEndElement();
        aWriter.WriteStartElement("Height");
        aWriter.WriteString(iHeight.ToString(nfi));
        aWriter.WriteEndElement();
        aWriter.WriteStartElement("AllowHits");
        aWriter.WriteString(iAllowHits.ToString());
        aWriter.WriteEndElement();
    }
    
    public override void Visit(Visitor aVisitor, bool aVisitNonActive) {
        aVisitor.AcceptHit(this);
        base.Visit(aVisitor, aVisitNonActive);
    }
    
    public float Width {
        get {
            return iWidth;
        }
        set {
            iWidth = value;
            OnSetWidth();
            ObserverUpdate();
        }
    }
    
    public virtual void OnSetWidth() {
    }
    
    public float Height {
        get {
            return iHeight;
        }
        set {
            iHeight = value;
            OnSetHeight();
            ObserverUpdate();
        }
    }
    
    public virtual void OnSetHeight() {
    }
    
    public bool AllowHits {
        get {
            return iAllowHits;
        }
        set {
            iAllowHits = value;
            ObserverUpdate();
        }
    }
    
    public override void Render() {
        Render(true);
    }
    
    public override void Render(bool aForce) {
        Rectangle r = new Rectangle((int)WorldSrt.Translation.X, (int)WorldSrt.Translation.Y, (int)Width, (int)Height);
        Renderer.Instance.Render(r, aForce);
    }
    
    public bool Inside(Vector3d aPosition) {
        Trace.WriteLine(Trace.kGui, "hitpos=" + aPosition + " w=" + iWidth + ", h=" + iHeight);
        if(aPosition.X >= 0 && aPosition.X < iWidth) {
            if(aPosition.Y >= 0 && aPosition.Y < iHeight) {
                //MsgHit msg = new MsgHit(Fullname, aPosition);
                //SendMessage(msg);
                return true;
            }
        }
        return false;
    }
    
    protected float iWidth = 0.0f;
    protected float iHeight = 0.0f;
    protected bool iAllowHits = true;
}
    
} // Scenegraph
} // Gui
} // Linn

File: Gui/Linn/Gui/SceneGraph/NodeHit.py
from Ft.Xml import XPath
import Linn.Gui.Serialize
import Linn.Gui.SceneGraph.Node

class NodeHit(Linn.Gui.SceneGraph.Node.Node):
    def __init__(self, aName=None, aWidth=0, aHeight=0):
        if aName == None:
            aName = self.UniqueName('NodeHit')
        Linn.Gui.SceneGraph.Node.Node.__init__(self, aName)
        self.iWidth = aWidth
        self.iHeight = aHeight
        
    def Load(self, aStream):
        Linn.Gui.SceneGraph.Node.Node.Load(self, aStream)
            
        width = aStream.LoadNode('Width')
        if len(width) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Width>', len(width), 1)
        self.iWidth = float(width[0].firstChild.nodeValue)
        #print 'width = ', self.iWidth
        
        height = aStream.LoadNode('Height')
        if len(height) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Height>', len(height), 1)
        self.iHeight = float(height[0].firstChild.nodeValue)
        #print 'height = ', self.iHeight
    
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'NodeHit')
            
        Linn.Gui.SceneGraph.Node.Node.Save(self, aStream)
        
        width = aStream.CreateElement('Width')
        width.appendChild(aStream.CreateTextNode(str(self.iWidth)))
        aStream.SaveNode(width)
        
        height = aStream.CreateElement('Height')
        height.appendChild(aStream.CreateTextNode(str(self.iHeight)))
        aStream.SaveNode(height)
    
    def Visit(self, aVisitor):
        Linn.Gui.SceneGraph.Node.Node.Visit(self, aVisitor)
        # accept visitor for this node
        aVisitor.AcceptHit(self)
            
    def Width(self):
        return self.iWidth
    
    def SetWidth(self, aWidth):
        self.iWidth = aWidth
        self.ObserverUpdate()
    
    def Height(self):
        return self.iHeight
    
    def SetHeight(self, aHeight):
        self.iHeight = aHeight
        self.ObserverUpdate()
            
    def IsInside(self, aPosition):
        if aPosition.iX > 0 and aPosition.iX < self.iWidth and \
           aPosition.iY > 0 and aPosition.iY < self.iHeight:
               return True
        else:
            return False

        
 
